pkg/controller/petset/pet_set_utils.go

```diff
@@ -0,0 +1,151 @@
+	"k8s.io/kubernetes/pkg/client/cache"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/golang/glog"
+)
+
+// overlappingPetSets sorts a list of PetSets by creation timestamp, using their names as a tie breaker.
+type overlappingPetSets []apps.PetSet
+
+func (o overlappingPetSets) Len() int      { return len(o) }
+func (o overlappingPetSets) Swap(i, j int) { o[i], o[j] = o[j], o[i] }
+
+func (o overlappingPetSets) Less(i, j int) bool {
+	if o[i].CreationTimestamp.Equal(o[j].CreationTimestamp) {
+		return o[i].Name < o[j].Name
+	}
+	return o[i].CreationTimestamp.Before(o[j].CreationTimestamp)
+}
+
+// updatePetCount attempts to update the Status.Replicas of the given PetSet, with a single GET/PUT retry.
+func updatePetCount(kubeClient *client.Client, ps apps.PetSet, numPets int) (updateErr error) {
+	if ps.Status.Replicas == numPets || kubeClient == nil {
+		return nil
+	}
+	psClient := kubeClient.Apps().PetSet(ps.Namespace)
+	var getErr error
+	for i, ps := 0, &ps; ; i++ {
+		glog.V(4).Infof(fmt.Sprintf("Updating replica count for PetSet: %s/%s, ", ps.Namespace, ps.Name) +
+			fmt.Sprintf("replicas %d->%d (need %d), ", ps.Status.Replicas, numPets, ps.Spec.Replicas))
+
+		ps.Status = apps.PetSetStatus{Replicas: numPets}
+		_, updateErr = psClient.UpdateStatus(ps)
+		if updateErr == nil || i >= statusUpdateRetries {
+			return updateErr
+		}
+		if ps, getErr = psClient.Get(ps.Name); getErr != nil {
+			return getErr
+		}
+	}
+}
+
+// claimClient returns the pvcClient for the given kubeClient/ns.
+func claimClient(kubeClient *client.Client, ns string) client.PersistentVolumeClaimInterface {
+	return kubeClient.PersistentVolumeClaims(ns)
+}
+
+// podClient returns the given podClient for the given kubeClient/ns.
+func podClient(kubeClient *client.Client, ns string) client.PodInterface {
+	return kubeClient.Pods(ns)
+}
+
+// unHealthyPetTracker tracks unhealthy pets for petsets.
+type unHealthyPetTracker struct {
+	store     cache.Store
+	storeLock sync.Mutex
+}
+
+// Get returns a previously recorded blocking pet for the given petset.
+func (u *unHealthyPetTracker) Get(ps *apps.PetSet, knownPets []*api.Pod) (*pcb, error) {
+	u.storeLock.Lock()
+	defer u.storeLock.Unlock()
+
+	// Retrieve by key but store by object because the store interface doesn't allow
+	// us to Get/Add a related obj (petset: blocking pet).
+	key, err := controller.KeyFunc(ps)
+	if err != nil {
+		return nil, err
+	}
+	obj, exists, err := u.store.GetByKey(key)
+	if !exists || err != nil {
+		return nil, err
+	}
+
+	blockingPet := obj.(*pcb)
+	hc := defaultPetHealthChecker{}
+	blockingPetName := fmt.Sprintf("%v/%v", blockingPet.pod.Namespace, blockingPet.pod.Name)
+
+	knownPetMap := map[string]*api.Pod{}
+	for _, pod := range knownPets {
+		knownPetMap[fmt.Sprintf("%v/%v", pod.Namespace, pod.Name)] = pod
+	}
+
+	// Healthy or deleted pods no longer block the petset
+	blockingPetPod, ok := knownPetMap[blockingPetName]
```

Comment:
fyi checking the store for status of blocking pet is racy w/ watch and I've fixed it locally, just so you don't stare too hard during review